hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/IncomingDataManagerImpl.java

```diff
@@ -0,0 +1,259 @@
+        try {
+            minReportingInterval = new Integer(
+                    AlertProperties.getProperty(MIN_REPORTING_INTERVAL, MIN_REPORTING_INTERVAL_DEFAULT));
+        } catch (Throwable t) {
+            if (log.isDebugEnabled()) {
+                t.printStackTrace();
+            }
+            log.errorf("Failed to initialize: %s", t.getMessage());
+        }
+    }
+
+    @Override
+    public void bufferData(IncomingData incomingData) {
+        executor.submit(() -> {
+            processData(incomingData);
+        });
+    }
+
+    @Override
+    public void bufferEvents(IncomingEvents incomingEvents) {
+        executor.submit(() -> {
+            processEvents(incomingEvents);
+        });
+    }
+
+    private void processData(IncomingData incomingData) {
+        log.debugf("Processing [%d] datums for AlertsEngine.", incomingData.incomingData.size());
+
+        // remove data not needed by the defined triggers
+        // remove duplicates and apply natural ordering
+        TreeSet<Data> filteredData = new TreeSet<Data>(filterIncomingData(incomingData));
+
+        // remove offenders of minReportingInterval
+        enforceMinReportingInterval(filteredData);
+
+        // check to see if any data can be used to generate data-driven group members
+        checkDataDrivenGroupTriggers(filteredData);
+
+        try {
+            log.debugf("Sending [%d] datums to AlertsEngine.", filteredData.size());
+            alertsEngine.sendData(filteredData);
+
+        } catch (Exception e) {
+            log.errorf("Failed to send [%d] datums:", filteredData.size(), e.getMessage());
+        }
+    }
+
+    private void processEvents(IncomingEvents incomingEvents) {
+        log.debugf("Processing [%d] events to AlertsEngine.", incomingEvents.incomingEvents.size());
+
+        // remove events not needed by the defined triggers
+        // remove duplicates and apply natural ordering
+        TreeSet<Event> filteredEvents = new TreeSet<Event>(filterIncomingEvents(incomingEvents));
+
+        // remove offenders of minReportingInterval
+        enforceMinReportingIntervalEvents(filteredEvents);
+
+        try {
+            alertsEngine.sendEvents(filteredEvents);
+        } catch (Exception e) {
+            log.errorf("Failed sending [%d] events: %s", filteredEvents.size(), e.getMessage());
+        }
+    }
+
+    private Collection<Data> filterIncomingData(IncomingData incomingData) {
+        Collection<Data> data = incomingData.getIncomingData();
+        data = incomingData.isRaw() ? dataIdCache.filterData(data) : data;
+
+        return data;
+    }
+
+    private Collection<Event> filterIncomingEvents(IncomingEvents incomingEvents) {
+        Collection<Event> events = incomingEvents.getIncomingEvents();
+        events = incomingEvents.isRaw() ? dataIdCache.filterEvents(events) : events;
+
+        return events;
+    }
+
+    private void enforceMinReportingInterval(TreeSet<Data> orderedData) {
```

Comment:
I am testing this one.
I think there is something not working fine.
After the fix proposed, the cluster is working but the enforceMinReportingInterval is not filtering as expected.
I am debugging it, I will post the insights.